[jerrie.choi] WAS 미션 Step2 PR

kakao-jerrie-choi opened the pull request:
안녕하세요 요한,
휴가는 잘 다녀오셨을까요?

이번이 마지막 미션입니다, 힘들었는데 막상 끝나려하니 아쉽네요 😥
몇 주간 고생 많으셨습니다! 덕분에 본격적인 업무 전 많은 걸 배우고 시작할 수 있게 되었습니다, 감사합니다 👏👏

이번 미션은 쿠키와 세션을 사용해 로그인 판별을 할 수 있도록 구현하는 것이 주된 요구사항이었습니다.

로그인을 성공하게 되면
- 쿠키를 생성하여 JSESSIONID와 Path를 응답
- 자바 어플리케이션 내부에서는 JSESSIONID를 키로 하는 세션을 생성해 그 안에 유저 정보를 저장
- 해당 세션은 세션 매니저에 저장

하는 과정을 거칩니다

그리고 로그인이 필요한 요청이 들어오게되면
- 요청 객체에서 Cookie를 추출하고, 해당 쿠키 안에 JSESSIONID를 추출
- JSESSIONID를 세션 매니저에서 검색하여 세션 조회
- 그러한 세션이 존재한다면 로그인 확인, 존재하지 않는다면 유효하지 않은 쿠키로 로그인하지 않은 것으로 간주

하는 과정으로 로그인 여부를 판별했습니다.

마지막까지 감사합니다 :)

HQkim committed:
docs: README.md 미션 설명, 기능 요구사항 추가

HQkim committed:
feat: "GET /index.html" 요청에 대한 응답 구현

HQkim committed:
test: "GET /css/styles.css" 요청에 대한 응답 테스트 추가

HQkim committed:
feat: "GET /css/styles.css" 요청에 대한 응답 테스트 추가

HQkim committed:
test: 회원가입에 요청에 대한 응답 테스트 추가

HQkim committed:
feat: 회원가입에 요청에 대한 응답 구현

HQkim committed:
test: 회원가입을 POST 메서드로 요청했을 때 응답 테스트 추가

HQkim committed:
feat: 회원가입을 POST 메서드로 요청했을 때 응답 구현

HQkim committed:
fix: 회원가입을 완료한 경우 index.html로 리다이렉트 되도록 수정

- GET 방식으로 회원가입을 요청하는 기능 및 테스트 삭제

HQkim committed:
refactor: Http 요청을 HttpRequest 객체에 담아서 처리하도록 변경

HQkim committed:
refactor: FrontController 패턴으로 구조 변경

kakao-jerrie-choi committed:
refac: 컨트롤러에서 사용할 어노테이션 추가

kakao-jerrie-choi committed:
refac: 불필요한 변수 제거

kakao-jerrie-choi committed:
feat: Http 관련 Map 구조를 모두 일급 컬렉션으로 전환

kakao-jerrie-choi committed:
refac: HttpUtils 메서드 분리, 일급 컬렉션 적용, 커스텀 예외 추가

kakao-jerrie-choi committed:
refac: HttpRequest에 일급 컬렉션 적용

kakao-jerrie-choi committed:
refac: HttpResponse 패키지 이동 및 빌더 패턴 적용

kakao-jerrie-choi committed:
refac: HttpMethod 패키지 이동

kakao-jerrie-choi committed:
feat: HttpStatus 커스텀 이넘 추가

kakao-jerrie-choi committed:
refac: 요청 객체를 controller에서 원하는 타입으로 매핑

- RequestBody와 RequestParams 어노테이션 사용 시에만 적용
- 커스텀 예외(UnsupportedRequestException) 추가

kakao-jerrie-choi committed:
feat: 커스텀 예외 추가

kakao-jerrie-choi committed:
refac: 바뀐 인자 타입과 반환 타입에 맞게 ViewController 수정

kakao-jerrie-choi committed:
refac: 바뀐 인자 타입과 반환 타입에 맞게 UserController 수정 및 Get 요청 추가

kakao-jerrie-choi committed:
test: get 요청 테스트 추가

kakao-jerrie-choi committed:
refac: 커스텀 예외 추가

kakao-jerrie-choi committed:
refac: 알맞은 메서드를 찾는 책임을 Controller로 이동

- ViewController에서 메서드 찾는 로직을 커스텀하기 위함

kakao-jerrie-choi committed:
feat: 일괄적으로 url과 파일을 매핑해주는 메서드 생성

- 매칭되지 않는다면 404

kakao-jerrie-choi committed:
feat: ViewController에서 적절한 메서드 찾는 기능 커스텀

kakao-jerrie-choi committed:
refac: RequestLine 도메인 생성

kakao-jerrie-choi committed:
refac: RequestLine으로 수정

kakao-jerrie-choi committed:
refac: RequestLine에 알맞게 메서드 수정

kakao-jerrie-choi committed:
Merge branch 'step2' into step1

kakao-jerrie-choi committed:
Merge pull request #1 from CHOIJXWXN/step1

Refactor

kakao-jerrie-choi committed:
docs: step2 기능 요구사항 추가

kakao-jerrie-choi committed:
feat: 유저를 찾을 수 없을 때 커스텀 예외 추가

kakao-jerrie-choi committed:
refac: 데이터베이스에서 유저 검색 시 Optional로 처리

kakao-jerrie-choi committed:
feat: 유저 패스워드 검사 기능 추가

kakao-jerrie-choi committed:
feat: FrontController에 로그인 url 등록

kakao-jerrie-choi committed:
feat: 로그인 인증 및 쿠키 저장

kakao-jerrie-choi committed:
feat: 쿠키 객체 생성

kakao-jerrie-choi committed:
feat: 쿠키 저장 여부로 로그인 판단 기능 추가

kakao-jerrie-choi committed:
feat: handlebars를 활용하여 동적 html로 변경

kakao-jerrie-choi committed:
feat: RequestHeader 어노테이션 추가

kakao-jerrie-choi committed:
feat: 유저 목록 검색 기능 추가

kakao-jerrie-choi committed:
feat: 유저 검색 url Front에 등록 및 매핑

kakao-jerrie-choi committed:
test: 로그인 테스트 코드 작성

kakao-jerrie-choi committed:
refac: response의 body에 빈 문자열 추가

kakao-jerrie-choi committed:
test: 유저 목록 조회 테스트 코드 작성

kakao-jerrie-choi committed:
refac: test에서 제대로 된 cookie 요청

kakao-jerrie-choi committed:
feat: HttpSession 구현

kakao-jerrie-choi committed:
feat: SessionManager 구현

kakao-jerrie-choi committed:
refac: 쿠키에 jsessionid getter 추가

kakao-jerrie-choi committed:
feat: 로그인 시 세션 저장 로직 추가

kakao-jerrie-choi committed:
refac: 로그인 판별 여부를 세션 검증으로 구체화

kakao-jerrie-choi committed:
fix: typo 수정

kakao-jerrie-choi committed:
refac: 매직스트링 상수화

leeyohan93 commented on src/main/java/controller/BaseController.java:
```diff
@@ -1,4 +1,19 @@
+
 public abstract class BaseController {
+
+    protected Method getProperMethod(CustomHttpRequest request) throws NoSuchMethodException {
```
현재 상속을 통해 자식 클래스에서 super 참조를 통해 부모 클래스의 메서드들을 직접 호출하고 있네요!
이러한 경우 두 클래스는 강하게 결합되기 때문에 합성을 사용하는 방법과 같이 super 호출을 제거해 
결합도를 제거할 수 있는 방법을 고민해보시면 좋을 것 같습니다 :)

오브젝트 책 내용 중 상속에 관한 내용을 요약한 링크 공유드릴게요 :)
https://incheol-jung.gitbook.io/docs/study/object/2020-03-10-object-chap10

추가적으로 FrontController 외에 getProperMethod()를 사용하는 곳이 없는 것 같아요!
이 부분도 함께 고민해보시면 좋을 것 같습니다!

leeyohan93 commented on src/main/java/utils/HttpUtils.java:
```diff
@@ -14,29 +13,14 @@
-        }
-
-        String protocol = tokens[2];
+        CustomHttpRequestLine httpRequestLine = new CustomHttpRequestLine(line);
```
CustomHttpRequestLine 객체를 만들어 책임을 이동시켜주셨네요 💯 💯 
이번에는 Utils 클래스의 존재 목적에 대해 고민해보셔도 좋을 것 같습니다!

제리는 유틸 클래스를 어떤 기준으로 사용하시나요? 메서드로 대체될 수는 없을까요?

kakao-jerrie-choi replied:
HttpUtils는 지금 요청을 읽고 이를 내부에서 사용할 수 있는 형태인 CustomHttpRequest로 가공해줍니다.
그리고, 응답 객체 CustomHttpResponse를 받아 외부에서 원하는 형태로 바꾸어 DataOutputStream으로 데이터를 전송해줍니다.

이러한 역할은 RequestHandler가 하는 게 적당할까요?
현재 RequestHandler로 메서드 형태로 옮긴다면 이제 사실상 핸들러가 아닌 거 같은데, 어떤 네이밍이 좋을까요

leeyohan93 replied:
핸들러가 아니라고 생각하신 기준도 궁금합니다 :) 핸들러는 어떤 기능을 해야한다는 기준이 있으실까요? 
메서드 형태로 옮기는 것이 부자연스러워 보인다면, Utils 클래스를 제거한다고 생각해보면 좋을 것 같아요. 
저는 개인적으로 이러한 객체 생성에서는 ~Mapper 클래스를 만들어 의존성 주입 받아 사용하거나, 정적 팩터리 메서드를 사용하는 편입니다!

leeyohan93 commented on src/main/java/model/http/SessionManager.java:
```diff
@@ -0,0 +1,30 @@
+
+    public Optional<CustomHttpSession> findSession(final String id) {
+        return Optional.of(SESSIONS.getOrDefault(id, null));
+    }
```
사소하지만 getOrDefault 대신 get을 사용해도 괜찮을 것 같아요 :)

```suggestion
    public Optional<CustomHttpSession> findSession(final String id) {
        return Optional.of(SESSIONS.get(id));
    }
```

leeyohan93 commented on src/main/java/model/http/CustomHttpSession.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class CustomHttpSession {
```
세션과 관련된 코드들에 테스트 코드가 함께 있었어도 좋았을 것 같아요 :)

leeyohan93 commented on src/main/java/controller/UserController.java:
```diff
@@ -38,4 +44,57 @@ public CustomHttpResponse createGET(@CustomRequestParams User user) {
+        }
+        headers.put("Content-Type", "text/html;charset=utf-8");
+
+        TemplateLoader loader = new ClassPathTemplateLoader();
+        loader.setPrefix("/templates");
+        loader.setSuffix(".html");
+        Handlebars handlebars = new Handlebars(loader);
+
+        Template template = handlebars.compile("user/list");
+        Collection<User> users = DataBase.findAll();
+        String page = template.apply(users);
```
템플릿에 대한 책임은 UserController 보다 적합한 위치가 필요해보이는 것 같아요 :)
좀 더 적절한 곳에 이동하거나 새로운 객체를 만드는 것도 고려해보시면 좋을 것 같습니다!

leeyohan93 commented on src/main/java/model/http/CustomHttpCookie.java:
```diff
@@ -0,0 +1,28 @@
+        this.path = "/";
+    }
+
+    public String createJSessionId() {
+        UUID uuid = UUID.randomUUID();
+        return uuid.toString();
+    }
```
접근 제어자를 Private 으로 설정하는 것이 좋을 것 같아요~

leeyohan93 reviewed (changes requested):
제리 안녕하세요~
이전 피드백을 잘 반영해주시고 2단계도 잘 구현해주셨네요 👍 👍 
저도 몇 주동안 제리와 함께 이야기하면서 많은 것을 배울 수 있는 좋은 시간이었습니다.
몇가지 소소한 코멘트들을 남겨두었으니 미션을 마무리하기 전에 고민해보시면 좋을 것 같아요! :)

kakao-jerrie-choi committed:
refac: jsessionid 생성 메서드 접근 제어자 수정

kakao-jerrie-choi committed:
refac: Service 레이어를 생성하여 비즈니스 로직 할당

kakao-jerrie-choi committed:
refac: Map 메서드 중 getOrDefault -> get 로 교체

kakao-jerrie-choi committed:
refac: BaseController 인터페이스화 및 getProperMethod 책임 이동

- BaseController를 인터페이스로 만들고 getProperMethod 삭제
- getProperMethod를 FrontController에서 구현
    - 뷰 컨트롤러의 경우 특수하므로 분기처리하여 따로 구현

kakao-jerrie-choi committed:
refac: HttpUtils 삭제 및 CustomHttpRequestFactory 생성

- 인풋값을 HttpRequest로 가공하는 역할을 Factory가 담당
- 그 외 아웃풋을 내는 메서드는 RequestHandler로 이동
- HttpUtils에서 상수를 import하던 HttpRequestLine import문 수정

kakao-jerrie-choi committed:
refac: 가독성을 위해 test 코드에 display name 추가

kakao-jerrie-choi committed:
test: 세션 저장 정보 확인 테스트 추가

leeyohan93 commented on src/main/java/controller/FrontController.java:
```diff
@@ -24,33 +27,54 @@ public class FrontController {
+    }
+
+    private Method getProperMethod(BaseController controller, CustomHttpRequest request) throws NoSuchMethodException {
+        if (controller instanceof ViewController) {
```
instanceof 를 통해 구체 클레스를 확인하는 방법 보다는 메시지를 던지는 방법을 고려해보시면 좋을 것 같아요 :)
만약 또 다른 컨트롤러가 생긴다면 해당 메서드에는 if 문이 반복해서 생기게 되지 않을까요?

leeyohan93 commented on src/main/java/service/UserService.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    public String getUserListHtml() throws IOException {
+        TemplateLoader loader = new ClassPathTemplateLoader();
+        loader.setPrefix("/templates");
+        loader.setSuffix(".html");
+        Handlebars handlebars = new Handlebars(loader);
+
+        Template template = handlebars.compile("user/list");
```
템플릿에 대한 부분은 UserService 보다는 ViewController와 같이 스프링의 ViewResolver 역할을 하는 곳에서 구현해보는건 어떨까요? 유저 서비스는 오직 비즈니스 로직만 신경쓰도록 하면 좋을 것 같습니다 :)

leeyohan93 commented on src/main/java/model/http/CustomHttpRequestFactory.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
+public class CustomHttpRequestFactory {
```
요청 객체를 만드는 별도의 클래스를 만들어주셨네요 👍 👍 
유틸클래스를 사용하는 것보다 더 명확하고 좋은 것 같습니다 :)

leeyohan93 reviewed (approved):
마지막까지 고생 많으셨습니다 🙏 
사소한 코멘트 남겨두었고 긴 시간동안 미션을 진행하면서 저도 함께 많이 배웠습니다. 😄 
시간이 여유가 되신다면 미션 하셨던 것들을 한번 천천히 돌아보아도 좋을 것 같아요.
수고하셨습니다 :)
